pkg/rpcclient/wsclient.go

```diff
@@ -542,16 +605,30 @@ readloop:
 	c.respLock.Unlock()
 	c.subscriptionsLock.Lock()
 	for rcvrCh, ids := range c.receivers {
-		rcvr := c.subscriptions[ids[0]]
+		c.dropSubCh(rcvrCh, ids[0], true)
+	}
+	c.subscriptionsLock.Unlock()
+	c.Client.ctxCancel()
+}
+
+// dropSubCh closes corresponding subscriber's channel and removes it from the
+// receivers map. If the channel belongs to a naive subscriber then it will be
+// closed manually without call to Close(). The channel is still being kept in
+// the subscribers map as technically the server-side subscription still exists
+// and the user is responsible for unsubscription. This method must be called
+// under subscriptionsLock taken. It's the caller's duty to ensure dropSubCh
+// will be called once per channel, otherwise panic will occur.
+func (c *WSClient) dropSubCh(rcvrCh any, id string, ignoreCloseNotificationChannelIfFull bool) {
+	if ignoreCloseNotificationChannelIfFull || c.opts.CloseNotificationChannelIfFull {
+		rcvr := c.subscriptions[id]
 		_, ok := rcvr.(*naiveReceiver)
-		if !ok { // naiveReceiver uses c.Notifications that is about to be closed below.
-			c.subscriptions[ids[0]].Close()
+		if ok { // naiveReceiver uses c.Notifications that should be handled separately.
+			close(c.Notifications)
```

Comment:
Only one `naiveReceiver` since there is one `c.Notifications` channel in `c.receivers`?

Reply:
Yes. Is something wrong here?

Reply:
At first glance it looks like there is a potential for double `close()` invocation here, but then it should be OK.

Reply:
It is removed from the receivers list immediately after that, so it should be OK.